ipwb/replay.py

```diff
@@ -179,6 +180,107 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, datetimeK, rest = ln.split(maxsplit=2)
+
+    matchDegree = getMatchDegree(surt, datetime, surtk, datetimeK)
+    if matchDegree == MementoMatch.EXACTMATCH:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, datetimeK, rest = line.split(maxsplit=2)
```

Comment:
To handle malformed lines safely:

```suggestion
        try:
            surtk, datetimeK, rest = line.split(maxsplit=2)
        except ValueError as e:
            continue
```